src/main/scala/nodes/learning/LBFGS.scala

```diff
@@ -0,0 +1,244 @@
+
+    val initialWeights = DenseVector.zeros[Double](numFeatures * numClasses)
+
+    val states =
+      lbfgs.iterations(new CachedDiffFunction(costFun), initialWeights)
+
+    /**
+      * NOTE: lossSum and loss is computed using the weights from the previous iteration
+      * and regVal is the regularization value computed in the previous iteration as well.
+      */
+    var state = states.next()
+    while (states.hasNext) {
+      lossHistory += state.value
+      state = states.next()
+    }
+    lossHistory += state.value
+    val finalWeights = state.x.asDenseMatrix.reshape(numFeatures, numClasses)
+
+    val lossHistoryArray = lossHistory.result()
+
+    logInfo("LBFGS.runLBFGS finished. Last 10 losses %s".format(
+      lossHistoryArray.takeRight(10).mkString(", ")))
+
+    finalWeights
+  }
+
+  /**
+    * CostFun implements Breeze's DiffFunction[T], which returns the loss and gradient
+    * at a particular point (weights). It's used in Breeze's convex optimization routines.
+    */
+  private class CostFun[T <: Vector[Double]](
+      data: RDD[T],
+      labels: RDD[DenseVector[Double]],
+      gradient: Gradient[T],
+      regParam: Double,
+      numExamples: Long,
+      numFeatures: Int,
+      numClasses: Int,
+      weightsIncludeBias: Boolean) extends DiffFunction[DenseVector[Double]] {
+
+    override def calculate(weights: DenseVector[Double]): (Double, DenseVector[Double]) = {
+      val weightsMat = weights.asDenseMatrix.reshape(numFeatures, numClasses)
+      // Have a local copy to avoid the serialization of CostFun object which is not serializable.
+      val bcW = data.context.broadcast(weightsMat)
+      val localGradient = gradient
+      val localNumFeatures = numFeatures
+      val localNumClasses = numClasses
+
+      val partitionGradientLosses = data.zipPartitions(labels) {
+        (partitionFeatures, partitionLabels) =>
+          Iterator.single(localGradient.compute(localNumFeatures, localNumClasses, partitionFeatures, partitionLabels, bcW.value))
+      }
+
+      val (gradientSum, lossSum) = MLMatrixUtils.treeReduce(
+        partitionGradientLosses,
+        (a: (DenseMatrix[Double], Double), b: (DenseMatrix[Double], Double)) => {
+          a._1 += b._1
+          (a._1, a._2 + b._2)
+        }
+      )
+
+      // total loss = lossSum / nTrain + 1/2 * lambda * norm(W)^2
+      val normWSquared = if (weightsIncludeBias) {
+        val weightsMinusBias = weightsMat(0 until (numFeatures - 1), ::)
+        val weightsNorm: Double = norm(weightsMinusBias.flatten())
+        weightsNorm * weightsNorm
+      } else {
+        math.pow(norm(weights), 2)
+      }
+      val regVal = 0.5 * regParam * normWSquared
+      val loss = lossSum / numExamples + regVal
+
+      // total gradient = gradSum / nTrain + lambda * w
+      val gradientTotal = gradientSum / numExamples.toDouble + (weightsMat * regParam)
+
+      (loss, gradientTotal.toDenseVector)
+    }
+  }
+}
+
+/**
+  * :: DeveloperApi ::
```

Comment:
We should be careful not to carry over artifacts from the spark codebase like this - we don't have any notion of DeveloperApi in keystone.